### ISSUE TITLE
[FIX-005] Reject unsupported topology/operations before claiming function-preserving widening

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#204]
**Category:** Model transformation integrity · **Priority:** P1 · **Size:** M

**Evidence status:** Reproduced preservation-contract defect.

## Outcome and scope

Prevent `widen` from claiming function preservation for an unproven execution path. The bounded fix supports known Sequential/FeedFwdNN connections and rejects unsupported width-dependent operations, arbitrary branches and shared aliases before replacement. Do not add general graph tracing, normalization transforms or new model recipe APIs.

## Current behavior / evidence

**Context and reproduction:** Seed torch with 4; create `Sequential(Linear(4,4), Softmax(dim=-1), Linear(4,2))`; call `widen(..., layer_name='0', new_width=7)` and compare both models on `torch.randn(3,4)`. Both forwards succeed, but max output error is **0.1229876429**. The width-dependent Softmax denominator changes when units are duplicated. The helper simply chooses the next registered Linear and silently accepts all intermediate modules. LayerNorm/BatchNorm width dependence or unrelated branch registration can also invalidate this inference.

## Repository map

- [`src/nnx/surgery/widen.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/surgery/widen.py#L33) — Copies the model, duplicates target rows, rescales downstream columns.
- [`src/nnx/surgery/widen.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/surgery/widen.py#L160) — Registration order is currently mistaken for data-flow evidence.
- [`src/nnx/nn/net/feed_fwd_nn.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/net/feed_fwd_nn.py#L23) — Actual per-layer activation/dropout execution must inform validation.
- [`src/nnx/nn/params/nn_params.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_params.py#L106) — Resolve effective per-layer operations instead of scalar defaults.
- [`tests/test_surgery_widen.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_surgery_widen.py#L24) — Supported-path parity and caller ownership tests.
- [`tests/test_surgery_widen.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_surgery_widen.py#L167) — Preserve ambient RNG independence.

## Implementation approach

1. Resolve and validate the target-to-consumer path before allocating replacements. Restrict automatic inference to a proven container execution order: ordinary `nn.Sequential`, or the explicit FeedFwdNN layer list and its effective per-layer operations. A generic module’s registration traversal is not enough, even when dimensions happen to match.
2. Initially allow direct Linear-to-Linear and Linear-to-ReLU/Identity-to-Linear paths. Reject Softmax, LayerNorm, BatchNorm, unrecognized operations, branching/aliased consumers and missing consumers with a `ValueError` naming the boundary. For FeedFwdNN resolve `activation_for(index)` and `dropout_for(index)`; reject an unsupported effective operation rather than guessing from the scalar defaults. Nonzero dropout cannot support a deterministic training-forward parity promise; document/restrict the supported mode accordingly.
3. Reuse the existing local-generator duplication and column-rescaling algorithm for validated paths. Retain no source mutation, no global RNG consumption, bias behavior, dtype/device and current missing-name/type errors. Coordinate trainability/mode preservation with FIX-016 without expanding this ticket into parameter-role policy.
4. Update `widen`’s docstring to state the actual supported topology. Any user guide changes belong in canonical `docs/surgery.md` through the existing docs projection, not hand-edited generated site/wiki files.

## Acceptance criteria

- [ ] The Softmax example fails before returning a transformed model, or a specifically implemented transform demonstrably preserves it.
- [ ] LayerNorm/BatchNorm and unrelated registered branches are rejected unless explicitly supported and tested.
- [ ] Existing direct Linear and Linear→ReLU→Linear widening retains numerical parity and RNG preservation.
- [ ] Validation failure leaves the source unchanged; the error names the unsupported boundary/operation.

## Validation and completion

Minimal rejection regression using the existing public API:

```python
import torch
from torch import nn
from nnx import widen

torch.manual_seed(4)
net = nn.Sequential(nn.Linear(4, 4), nn.Softmax(dim=-1), nn.Linear(4, 2))
snapshot = {k: v.clone() for k, v in net.state_dict().items()}
try:
    widen(net, layer_name="0", new_width=7)
except ValueError as exc:
    assert "Softmax" in str(exc) or "unsupported" in str(exc).lower()
else:
    raise AssertionError("unsupported Softmax boundary was accepted")
assert all(torch.equal(snapshot[k], v) for k, v in net.state_dict().items())
```

Proposed cases: `test_widen_rejects_width_dependent_intermediate`, `test_widen_rejects_unproven_registered_branch`, `test_widen_rejects_shared_alias`, and `test_widen_resolves_effective_feedforward_activation`. Parametrize Softmax/LayerNorm/BatchNorm; add a module whose registered second Linear is not the actual consumer. Preserve existing direct and ReLU parity (`atol=1e-5`), no-bias, deterministic seed and ambient RNG tests. On rejected paths compare original tensor values, module identities and RNG state. Never relax parity tolerances merely to accept a new topology.

From the repository root, use `uv sync --frozen --all-extras` for the CI-equivalent environment. Existing focused coverage (files already exist):

```bash
uv run --frozen pytest tests/test_surgery_widen.py tests/test_per_layer_act_dropout.py
```

Add the proposed regression cases above to those existing test files, then rerun the same command. Finish with:

```bash
uv run --frozen ruff check src/ tests/ examples/ scripts/
uv run --frozen ruff format --check src/ tests/ examples/ scripts/
uv run --frozen pyright --warnings
```

No new dependency is needed for the implementation described. Preserve Python 3.10 compatibility and default-field omission in existing serialization. These commands are completion requirements for implementation; this specification review does not claim they were run or passed.

## Cascading changes and affected consumers

- **Supported forward graph.** Validate using the actual [FeedFwdNN.forward](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/net/feed_fwd_nn.py) activation/dropout behavior, not only registration order in [_find_next_linear](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/surgery/widen.py). The stronger failure contract must reach every public facade and clearly name unsupported Softmax/norm/branch/alias boundaries. Direct and ReLU paths remain usable; evaluation parity with disabled dropout must not be presented as identical stochastic training. Preserve failure without source/RNG mutation before any caller rebuilds an optimizer or creates a run.

- **Reconstruction caller.** The [widen → refine guide](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/surgery.md) reconstructs a new NNModel around widened weights. Keep its new hidden width and all existing per-layer activation/dropout overrides aligned, preferably by replacing the immutable params instead of rebuilding only scalar fields. Do not assign a widened net beneath stale NNModel.net_params and then claim ordinary checkpoint reconstruction. Build a fresh correctly described model, load the widened state, train with a newly constructed optimizer, and test BEST reload. Exact resume from the pre-surgery optimizer topology is outside this fix; future recipe persistence remains [FEAT-016](https://github.com/thekaveh/NNx/issues/238).

- **Honest samples and claims.** The [surgery facade](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/surgery/__init__.py), README and guide currently promise broad/drop-in function preservation. Narrow those statements to validated topology and execution mode, and remove the claim that every structural edit is directly reconstructible through unchanged NNParams. Add a tiny widening branch to [the surgery example](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/20_low_rank_surgery_ffn.py) that demonstrates supported eval parity plus a rejected Softmax case before its independent low-rank section. Execute this branch under a bounded sample test, retaining the full teaching workload separately.

### Samples and executable coverage

- [`examples/20_low_rank_surgery_ffn.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/20_low_rank_surgery_ffn.py): Add a short independent widening contract helper with direct/ReLU parity, Softmax rejection and correctly described fresh NNModel reconstruction; leave low-rank manual refinement separate. **Verify:** Proposed execution node: uv run --frozen pytest tests/test_examples_smoke.py -k widen_supported_workflow. CPU, offline, temporary working directory; call the bounded helper rather than the full training main. Retain import-without-main coverage and add the callable to explicit execution coverage; import success alone is insufficient.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`docs/surgery.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/surgery.md) | `generated/site/Model-surgery.md` | `generated/wiki/Model-surgery.md` | Replace unconditional widening guarantees with supported topology/mode table; rebuild widened NNModel using copied full params including overrides and fresh optimizer, then reload BEST; remove universal drop-in persistence claim. |
| [`README.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/README.md) | `generated/site/Repository-Overview.md` | `generated/wiki/Repository-Overview.md` | Qualify surgery function-preservation and refinement claims to supported topology and explicit reconstruction; avoid implying arbitrary edits resume the old optimizer state. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] A supported FeedFwdNN widening workflow rebuilds NNModel with the new width and preserved per-layer overrides, completes a short fit and reloads BEST with matching pre-save predictions.
- [ ] Unsupported widening fails before a caller-created training run or optimizer replacement, leaving the original model and ambient RNG intact through the public nnx facade.
- [ ] The bounded surgery-example widening branch executes both a valid ReLU path and a rejected Softmax path; it distinguishes eval parity from stochastic training behavior.
- [ ] Surgery facade docstrings and the runnable guide no longer imply arbitrary module registration order or stale NNParams can guarantee function-preserving, reconstructible refinement.
- [ ] The issue-specific changes in `docs/surgery.md`, `README.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** Propagation belongs to this fix and existing APIs. Related tickets coordinate shared contracts/files; they are not prerequisites for the bounded repaired path.

Additional nonblocking consumers/coordination: [FIX-006](https://github.com/thekaveh/NNx/issues/205), [FIX-016](https://github.com/thekaveh/NNx/issues/226), [FEAT-016](https://github.com/thekaveh/NNx/issues/238).

## Delivery order and dependencies

- **Execution order:** 05 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.
- **Related / optional integration:** [FEAT-016](https://github.com/thekaveh/NNx/issues/238). These links do not block this ticket.

**Dependency rationale.** Independent repair of current behavior. No dependency; FEAT-016 must respect the supported topology boundary.

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved.

**Root cause:** `widen` picked the next `nn.Linear` in `named_modules()` registration order and accepted every module in between, so registration order was mistaken for data flow. A `Linear → Softmax → Linear` net was "widened" with a 0.12 output drift (Softmax's denominator is width-dependent), LayerNorm/BatchNorm boundaries produced models that crashed on the next forward, and a registered-but-unconnected branch or an aliased module could be rescaled as if it were the consumer.

**What changed**
- New `_resolve_consumer` validates the target→consumer path on the **source** before anything is copied or allocated. The consumer is inferred only where the execution order is proven: a target directly inside an `nn.Sequential` (following siblings walked in registration order) or in `FeedFwdNN.layers` (the *effective* `activation_for(i)` / `dropout_for(i)` resolved, not the scalar defaults). Every op between target and consumer must be elementwise (`ReLU, LeakyReLU, GELU, Tanh, Sigmoid, ELU, SELU, Softplus, SiLU, Identity, Dropout`; `Activations` members except `SOFTMAX`). Width-dependent ops, other containers, missing consumers and modules registered under more than one path (aliases) raise a `ValueError` naming the boundary; the source model, its module identities and the RNG are untouched. Duplication/rescaling, local generator, bias, dtype/device and the FIX-016 role/mode copy are unchanged; parity through elementwise ops is covered for ten module types (eval-mode parity when dropout is configured, documented).
- Example 20: bounded `widen_supported_workflow()` (`tests/test_examples_smoke.py -k widen_supported_workflow`) — ReLU parity, Softmax rejection with the source untouched, and the honest reconstruction path: widen → `dataclasses.replace(net_params, hidden_dims=…)` (per-layer overrides preserved) → fresh `NNModel` → load state → parity → one epoch → BEST reload.
- Docs: `surgery.md` §1 table cell, §2 rebuild snippet now uses `replace(...)` instead of a hand-built `NNParams` that silently dropped per-layer overrides, §5 new bullet; README (+ PyPI readme) narrows the drop-in function-preservation claim; surgery facade docstring; `concepts.md` surgery bullet; examples catalog; CHANGELOG; API reference.

**PRs:** #291 (develop, merged), #292 (main, merged).

**Checks run (both branches):**
- `uv run --frozen pytest -q` → 1720 passed, 1 skipped, 2 deselected (the two libcairo-dependent `test_docs_projection` tests — pre-existing local-env failures, unrelated; CI runs them and passed). New: `test_widen_rejects_width_dependent_intermediate[softmax|layernorm|batchnorm]` (values, module identities and RNG state untouched), `test_widen_accepts_elementwise_intermediates[…×10]`, `test_widen_rejects_unproven_registered_branch`, `test_widen_rejects_missing_consumer`, `test_widen_rejects_shared_alias` (target and consumer aliases), `test_widen_resolves_effective_feedforward_activation`; existing direct/ReLU/no-bias parity, seed-determinism and ambient-RNG tests unchanged.
- `ruff check` → All checks passed; `ruff format --check` → all formatted; `pyright --warnings` → 0 errors, 0 warnings
- docs chain (`build_api_reference` / `build_docs` / `mkdocs build --strict` / `build_wiki --check` / `build_pypi_readme --check` / projection tests) → pass; `python examples/20_low_rank_surgery_ffn.py` runs end to end.
- CI: lint-and-test on Python 3.10–3.14, pip-audit, CodeQL, docs build all green on #291 and #292.

**Coordination:** FIX-016 (#226, already merged) supplied the role/mode copy this fix builds on; recipe persistence stays with FEAT-016. Live Pages/wiki publication is left to the `docs.yml` main-push job.
